### PR TITLE
[android][local-authentication] Fixed device credentials fallback when biometric sensors are unavailable

### PR DESCRIPTION
**Pull request by hubastard**
# Why

On Android when `disableDeviceFallback` is `false` and the device fingerprint sensor returns a `not_available` error, the fallback to Device Credentials does not work.

# How

To fix this bug, whenever the options includes `disableDeviceFallback=false` and we get one of the following error: 
`ERROR_HW_NOT_PRESENT`, `ERROR_HW_UNAVAILABLE`,`ERROR_NO_BIOMETRICS`, `ERROR_NO_SPACE` or `ERROR_UNABLE_TO_PROCESS`
We then manually prompt the device credentials unlock screen either using `BiometricManager.Authenticators.DEVICE_CREDENTIAL` for Android SDK 30+ or `KeyguardManager` for older versions.

# Test Plan

Since there isn't an easy way to repro these error codes on an emulator, I tested my changes by adding the `BiometricPrompt.ERROR_USER_CANCELED` to the list of codes in `isBiometricUnavailable()` and with the bare-expo by cancelling the biometric prompt I was able to make sure that the device credentials fallback was working properly.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt [hunk 1]
```diff
@@ -245,11 +283,80 @@ class LocalAuthenticationModule(context: Context) : ExportedModule(context), Act
...
+  fun promptDeviceCredentialsFallback(options: Map<String?, Any?>, promise: Promise) {
+    val fragmentActivity = currentActivity as FragmentActivity?
+    if (fragmentActivity == null) {
+      promise.resolve(
+              Bundle().apply {
+                putBoolean("success", false)
+                putString("error", "not_available")
+                putString("warning", "getCurrentActivity() returned null")
+              }
+      )
```
Comment (lukmccall) [inline packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt:295]:
```suggestion
      promise.resolve(
        Bundle().apply {
          putBoolean("success", false)
          putString("error", "not_available")
          putString("warning", "getCurrentActivity() returned null")
        }
      )
```

File: packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt [hunk 2]
```diff
@@ -245,11 +283,80 @@ class LocalAuthenticationModule(context: Context) : ExportedModule(context), Act
...
+      Runnable {
+        // On Android device older than 11, we need to use Keyguard to unlock by Device Credentials.
+        if (Build.VERSION.SDK_INT < 30) {
+          val credentialConfirmationIntent = keyguardManager.createConfirmDeviceCredentialIntent(promptMessage, "")
+          fragmentActivity.startActivityForResult(credentialConfirmationIntent, DEVICE_CREDENTIAL_FALLBACK_CODE);
+          return@Runnable
+        }
+
+        val executor: Executor = Executors.newSingleThreadExecutor()
+        biometricPrompt = BiometricPrompt(fragmentActivity, executor, authenticationCallback)
```
Comment (lukmccall) [inline packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt:314]:
```suggestion
	val localBiometricPrompt = BiometricPrompt(fragmentActivity, executor, authenticationCallback)
	if (localBiometricPrompt == null) {
	  promise.reject("E_INTERNAL_ERRROR", "Canceled authentication due to an internal error")
	  return@Runnable
	}
	biometricPrompt = localBiometricPrompt
```

File: packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt [hunk 3]
```diff
@@ -245,11 +283,80 @@ class LocalAuthenticationModule(context: Context) : ExportedModule(context), Act
...
+          promptInfoBuilder.setTitle(it)
+        }
+        promptInfoBuilder.setAllowedAuthenticators(BiometricManager.Authenticators.DEVICE_CREDENTIAL)
+        promptInfoBuilder.setConfirmationRequired(requireConfirmation)
+        val promptInfo = promptInfoBuilder.build()
+        try {
+          biometricPrompt!!.authenticate(promptInfo)
+        } catch (ex: NullPointerException) {
+          promise.reject("E_INTERNAL_ERRROR", "Canceled authentication due to an internal error")
+        }
```
Comment (lukmccall) [inline packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt:326]:
```suggestion
        localBiometricPrompt.authenticate(promptInfo)
```

File: packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt [hunk 4]
```diff
@@ -245,11 +283,80 @@ class LocalAuthenticationModule(context: Context) : ExportedModule(context), Act
...
+    // When Biometric is unavailable and using Keyguard fallback, the result will be handled here.
+    if (requestCode == DEVICE_CREDENTIAL_FALLBACK_CODE) {
+      if (resultCode == Activity.RESULT_OK) {
+        promise?.resolve(
+          Bundle().apply {
+            putBoolean("success", true)
+          }
+        )
+      } else {
+
```
Comment (lukmccall) [inline packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt:341]:
```suggestion
```

File: packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt [hunk 5]
```diff
@@ -245,11 +283,80 @@ class LocalAuthenticationModule(context: Context) : ExportedModule(context), Act
...
+    }
+
+    val promptMessage =  options["promptMessage"] as? String ?: ""
+    val requireConfirmation = options["requireConfirmation"] as? Boolean ?: true
+
+    // BiometricPrompt callbacks are invoked on the main thread so also run this there to avoid
+    // having to do locking.
+    uIManager.runOnUiQueueThread(
+      Runnable {
+        // On Android device older than 11, we need to use Keyguard to unlock by Device Credentials.
```
Comment (lukmccall) [inline packages/expo-local-authentication/android/src/main/java/expo/modules/localauthentication/LocalAuthenticationModule.kt:306]:
```suggestion
        // On Android devices older than 11, we need to use Keyguard to unlock by Device Credentials.
```

**Review by lukmccall — CHANGES_REQUESTED** [review]
Thanks for your contribution. Looking good. Please apply my suggestions.

**Review by lukmccall — APPROVED** [review]
Please, resolve the confit ;)
After that, I'll merge your PR.

**Comment by hubastard** [comment]
@lukmccall done

**Pull request merged**